File: sdks/node/src/executor.ts
import { clearSink, type ExecutorSink, installSink } from "./detached";
import type { Emitter } from "./events";
import type { Middleware } from "./middleware";
import {
  type NativeExecutor,
  type NativeQueue,
  startExecutor as startNativeExecutor,
} from "./native";
import type { ResourceRuntime } from "./resources";
import type { PayloadCodec, Serializer } from "./serializers";
import { createTaskCallback } from "./task-callback";
import type { RegisteredTask } from "./types";
import { createLogger } from "./utils";

const log = createLogger("executor");

/** How an executor attaches. Durations are milliseconds, per Node convention. */
export interface ExecutorRunOptions {
  /**
   * Scheduler address: `host:port`, `:port`, or `unix:/run/taskito.sock`.
   * Defaults to `$FLEXIQ_ATTACH`.
   */
  attach?: string;
  /** Jobs to run at once. Defaults to `$FLEXIQ_SLOTS`, then 1. */
  slots?: number;
  /** Shared secret, when the scheduler requires one. Defaults to `$FLEXIQ_ATTACH_TOKEN`. */
  token?: string;
  /** Identity announced to the scheduler. Defaults to one generated per process. */
  executorId?: string;
  /** How long to wait for the connection (default 10000). */
  connectTimeoutMs?: number;
  /** How often to send a liveness heartbeat (default 5000). */
  heartbeatIntervalMs?: number;
  /** How long a drain waits for in-flight jobs before disconnecting (default 30000). */
  shutdownDrainMs?: number;
  /** Only advertise these tasks. Defaults to every registered task. */
  tasks?: readonly string[];
}

/** Inputs assembled by {@link Queue.runExecutor}. @internal */
export interface ExecutorStartParams {
  /** Called once the executor has stopped, so the queue can forget it. */
  onStopped?: () => void;
  tasks: ReadonlyMap<string, RegisteredTask>;
  serializer: Serializer;
  codecs?: ReadonlyMap<string, PayloadCodec>;
  /**
   * The middleware chain for a task, minus `disabled`.
   *
   * Takes the disable list rather than reading one: an executor has no
   * settings store, so the scheduler resolves it and attaches it to each
   * dispatch.
   */
  middlewareFor: (taskName: string, disabled: readonly string[]) => readonly Middleware[];
  emitter: Emitter;
  resources: ResourceRuntime;
  run?: ExecutorRunOptions;
}

/**
 * A running attachment to a detached scheduler.
 *
 * The inverse of a {@link Worker}: instead of polling storage for work, it
 * dials a scheduler that already holds the database connection and runs
 * whatever it is sent. Task execution is identical — same middleware, codecs,
 * resources and cancel signal — because only the transport differs.
 */
export class Executor {
  private stopped?: Promise<void>;

  private constructor(
    private readonly native: NativeExecutor,
    /** The detached stand-in this attach answers for, and whose sink it holds. */
    private readonly queue: NativeQueue,
    private readonly sink: ExecutorSink,
    private readonly resources: ResourceRuntime,
    private readonly emitter: Emitter,
    private readonly onStopped?: () => void,
  ) {}

  /**
   * Attach and start running jobs. Use {@link Queue.runExecutor} rather than
   * calling this directly.
   *
   * @internal
   */
  static async start(queue: NativeQueue, params: ExecutorStartParams): Promise<Executor> {
    const { tasks, serializer, codecs, middlewareFor, emitter, resources, run, onStopped } = params;

    const address = run?.attach ?? process.env.FLEXIQ_ATTACH;
    if (!address) {
      throw new Error(
        "no scheduler address: pass `attach` or set FLEXIQ_ATTACH (e.g. scheduler:7777)",
      );
    }
    const advertised = [...(run?.tasks ?? tasks.keys())];

    // The executor does not exist yet, and the callback it needs must already
    // be able to reach it — a cancel frame lands in native state that a running
    // handler polls. Resolved through this holder, assigned once the attach
    // succeeds.
    //
    // The native attach starts its job loop before this promise resolves, so
    // the scheduler can dispatch into that window. An invocation waits for the
    // holder to be filled rather than reading it empty, which would run a
    // middleware the dispatch said was disabled and drop the job's progress.
    let attached: NativeExecutor | undefined;
    let markAttached: () => void = () => {};
    const attachedReady = new Promise<void>((resolve) => {
      markAttached = resolve;
    });

    const invoke = createTaskCallback({
      tasks,
      serializer,
      codecs,
      // Every one of these reaches for the executor rather than for storage,
      // which this process deliberately has none of: the scheduler holds the
      // connection and does the work on its behalf.
      middlewareFor: (taskName, jobId) =>
        middlewareFor(taskName, attached?.disabledMiddleware(jobId) ?? []),
      emitter,
      resources,
      queue,
      isCancelled: (jobId) => attached?.isCancelRequested(jobId) ?? false,
      // Overridden rather than left to `queue`, which is only the detached
      // stand-in when this process is a `taskito executor`. An executor started
      // from a process that *does* have storage would otherwise write progress
      // into its own database, where the job it names does not exist — the row
      // belongs to the scheduler. `queue`'s own route to the scheduler is the
      // sink installed below, for app code that calls the queue directly.
      setProgress: (jobId, progress) => attached?.reportProgress(jobId, progress),
      writeTaskLog: (jobId, taskName, level, message, extra) =>
        attached?.writeTaskLog(jobId, taskName, level, message, extra),
    });

    const taskCallback: typeof invoke = async (invocation) => {
      await attachedReady;
      return invoke(invocation);
    };

    const native = await startNativeExecutor(taskCallback, {
      address,
      tasks: advertised,
      slots: run?.slots ?? envInt("FLEXIQ_SLOTS"),
      // Env by preference for the token: in argv it shows up in `ps` output and
      // shell history.
      token: run?.token ?? process.env.FLEXIQ_ATTACH_TOKEN,
      executorId: run?.executorId,
      connectTimeoutMs: run?.connectTimeoutMs,
      heartbeatIntervalMs: run?.heartbeatIntervalMs,
      shutdownDrainMs: run?.shutdownDrainMs,
    });

    attached = native;
    const sink: ExecutorSink = {
      updateProgress: (jobId, progress) => native.reportProgress(jobId, progress),
      writeTaskLog: (jobId, taskName, level, message, extra) =>
        native.writeTaskLog(jobId, taskName, level, message, extra),
    };
    // Before `markAttached`, so the first invocation through the gate already
    // has somewhere to report to. Scoped to `queue`, so a second executor in
    // this process routes its own writes and neither steals the other's.
    installSink(queue, sink);
    if (!native.supportsSideChannel()) {
      log.warn(
        () =>
          `scheduler ${native.schedulerId} applies no progress or task logs on this executor's ` +
          "behalf; they will be dropped. Upgrade the scheduler to keep them.",
      );
    }
    markAttached();
    try {
      // Only lease the resource runtime once the attach actually succeeded, so a
      // refused handshake leaks nothing.
      resources.acquireWorker();
      emitter.emit("worker.started", { workerId: native.executorId });
    } catch (error) {
      // The session is live by now and no caller holds an `Executor` to stop
      // it, so a throwing resource factory or `worker.started` listener would
      // leak the attach until the process exits.
      await native.shutdown().catch((failure) => {
        log.debug(() => "releasing the attach after a failed start failed", failure);
      });
      throw error;
    }

    return new Executor(native, queue, sink, resources, emitter, onStopped);
  }

  /** Identity the scheduler announced when it accepted this attach. */
  get schedulerId(): string {
    return this.native.schedulerId;
  }

  /** Identity this executor attached under. */
  get executorId(): string {
    return this.native.executorId;
  }

  /** Peer label of the scheduler connection. */
  get peer(): string {
    return this.native.peer;
  }

  /** Whether the scheduler session is still open. */
  get running(): boolean {
    return this.native.isRunning();
  }

  /**
   * Resolve once the scheduler ends the session — a shutdown frame, or the
   * connection dropping. Does not drain; call {@link Executor.stop} for that.
   */
  wait(): Promise<void> {
    return this.native.wait();
  }

  /**
   * Drain in-flight work, disconnect, and release worker-scoped resources.
   *
   * Idempotent, and memoized like {@link Worker.stop} so a signal handler
   * racing a scheduler shutdown does not tear down twice.
   */
  stop(): Promise<void> {
    this.stopped ??= this.teardown();
    return this.stopped;
  }

  private async teardown(): Promise<void> {
    try {
      await this.native.shutdown();
    } finally {
      // The sink frames to a session that is over; leaving it installed would
      // turn a late report into a silent no-op instead of the warning it is.
      // Only this attach's own — another executor may have taken the slot since.
      clearSink(this.queue, this.sink);
      try {
        await this.resources.teardownWorker();
      } catch (error) {
        log.debug(() => "resource release during executor shutdown failed", error);
      }
      this.emitter.emit("worker.stopped", { workerId: this.executorId });
      this.onStopped?.();
    }
  }
}

/** Read a positive integer from the environment, or `undefined` when unusable. */
function envInt(name: string): number | undefined {
  const raw = process.env[name];
  if (raw === undefined || raw === "") {
    return undefined;
  }
  const value = Number(raw);
  if (!Number.isInteger(value) || value < 1) {
    throw new RangeError(`${name} must be a positive integer, got "${raw}"`);
  }
  return value;
}

File: sdks/java/src/main/java/org/byteveda/flexiq/worker/Executor.java
package org.byteveda.flexiq.worker;

import com.fasterxml.jackson.databind.ObjectMapper;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.function.Consumer;
import org.byteveda.flexiq.events.Emitter;
import org.byteveda.flexiq.events.EventName;
import org.byteveda.flexiq.events.FlexiQEvent;
import org.byteveda.flexiq.events.WorkerEvent;
import org.byteveda.flexiq.internal.JniExecutorControl;
import org.byteveda.flexiq.internal.NativeExecutor;
import org.byteveda.flexiq.logging.FlexiQLogger;
import org.byteveda.flexiq.middleware.Middleware;
import org.byteveda.flexiq.resources.ResourceRuntime;
import org.byteveda.flexiq.serialization.JsonSerializer;
import org.byteveda.flexiq.serialization.PayloadCodec;
import org.byteveda.flexiq.serialization.Serializer;
import org.jspecify.annotations.Nullable;

/**
 * Runs tasks for a detached scheduler instead of polling storage.
 *
 * <p>The inverse of {@link Worker}: the scheduler holds the database connection
 * and dispatches jobs over a socket, so this process runs task bodies without
 * any database credentials of its own. Handlers, middleware, codecs and the
 * cancel signal behave exactly as they do in a worker — only the transport
 * differs.
 *
 * <pre>{@code
 * try (Executor executor = Executor.builder()
 *         .discover()                 // handlers from META-INF/services
 *         .attach("scheduler:7777")
 *         .slots(4)
 *         .start()) {
 *     executor.awaitSession();        // until the scheduler shuts down
 * }
 * }</pre>
 */
public final class Executor implements AutoCloseable {
    private static final FlexiQLogger LOG = FlexiQLogger.create("executor");
    private static final int SHUTDOWN_TIMEOUT_SECONDS = 30;

    private final JniExecutorControl control;
    private final ExecutorService handlerPool;
    private final ResourceRuntime resources;
    private final Emitter emitter;
    private boolean closed;

    private Executor(
            JniExecutorControl control, ExecutorService handlerPool, ResourceRuntime resources, Emitter emitter) {
        this.control = control;
        this.handlerPool = handlerPool;
        this.resources = resources;
        this.emitter = emitter;
    }

    /**
     * A builder for an executor. Register handlers, point it at a scheduler, start.
     *
     * @return the builder
     */
    public static Builder builder() {
        return new Builder();
    }

    /**
     * Identity the scheduler announced when it accepted this attach.
     *
     * @return the scheduler's id
     */
    public String schedulerId() {
        return control.schedulerId();
    }

    /**
     * Identity this executor attached under.
     *
     * @return the id the scheduler knows this executor by
     */
    public String executorId() {
        return control.executorId();
    }

    /**
     * Peer label of the scheduler connection.
     *
     * @return the label, for logs and diagnostics
     */
    public String peer() {
        return control.peer();
    }

    /**
     * Whether this executor is still accepting work.
     *
     * @return {@code false} once the session ended or {@code stop()} was called
     */
    public boolean isRunning() {
        return control.isRunning();
    }

    /**
     * Block until this executor stops accepting work — the scheduler ending the
     * session, or a local {@link #stop()}. Does not drain; {@link #close()} does.
     */
    public void awaitSession() {
        control.awaitSession();
    }

    /**
     * Ask the scheduler to stop sending work and finish what is in flight.
     * Returns at once, so it is safe from a shutdown hook.
     */
    public void stop() {
        control.stop();
    }

    /**
     * Drain in-flight work, disconnect, and release handler threads.
     *
     * <p>The handler pool is drained before the native handle is freed: a handler
     * may still be completing its job through {@code control}, and the handle has
     * to outlive every such call. Idempotent.
     */
    @Override
    public void close() {
        if (closed) {
            return;
        }
        closed = true;
        control.stop();
        handlerPool.shutdown(); // stop accepting; let running handlers finish
        try {
            if (!handlerPool.awaitTermination(SHUTDOWN_TIMEOUT_SECONDS, TimeUnit.SECONDS)) {
                handlerPool.shutdownNow();
                if (!handlerPool.awaitTermination(SHUTDOWN_TIMEOUT_SECONDS, TimeUnit.SECONDS)) {
                    LOG.warn("handler threads still running after " + (2 * SHUTDOWN_TIMEOUT_SECONDS)
                            + "s; closing the executor handle anyway");
                }
            }
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
        try {
            control.close();
        } finally {
            resources.teardownWorker();
            emitter.emit(new WorkerEvent(EventName.WORKER_STOPPED, List.of()));
        }
    }

    /** Registers handlers and attach options, then starts the executor. */
    public static final class Builder {
        private static final ObjectMapper JSON = new ObjectMapper();

        private final Map<String, RegisteredTask> handlers = new HashMap<>();
        private final Map<EventName, List<Consumer<FlexiQEvent>>> listeners = new LinkedHashMap<>();
        private Serializer serializer = new JsonSerializer();
        private List<Middleware> middleware = List.of();
        private Map<String, PayloadCodec> codecs = Map.of();
        private ResourceRuntime resources = new ResourceRuntime();
        private @Nullable String address;
        private @Nullable String token;
        private @Nullable String executorId;
        private int slots = 1;
        private @Nullable Long connectTimeoutMs;
        private @Nullable Long heartbeatIntervalMs;
        private @Nullable Long shutdownDrainMs;

        private Builder() {}

        /**
         * Register every handler discoverable on the classpath.
         *
         * <p>The {@code @TaskHandler} processor lists a provider per annotated
         * class in {@code META-INF/services}, so this needs no application code
         * at all. A class the executor cannot construct is skipped at build time
         * with a compiler note — register those explicitly.
         *
         * <p>Discovery never replaces a handler already registered on this builder;
         * call {@code register(...)} <em>after</em> it to override one deliberately.
         *
         * @return {@code this}, for chaining
         */
        public Builder discover() {
            return discover(HandlerDiscovery.contextLoader());
        }

        /**
         * {@link #discover()} against a specific class loader.
         *
         * @param loader where to look for {@code META-INF/services} entries
         * @return {@code this}, for chaining
         */
        public Builder discover(ClassLoader loader) {
            HandlerDiscovery.load(loader, handlers.keySet(), "executor").forEach(this::register);
            return this;
        }

        /**
         * Register a generated {@code <Class>Tasks.handlers(impl)} bundle.
         *
         * @param registry the bundle whose handlers to take on
         * @return {@code this}, for chaining
         */
        public Builder register(HandlerRegistry registry) {
            registry.handlers().forEach(this::register);
            return this;
        }

        /**
         * Register one handler.
         *
         * @param handler the task descriptor paired with the code that runs it
         * @return {@code this}, for chaining
         */
        public Builder register(Handler<?, ?> handler) {
            handlers.put(
                    handler.task().name(),
                    new RegisteredTask(
                            handler.task().payloadType(),
                            cast(handler.function()),
                            handler.task().codecNames(),
                            handler.task().retryOn()));
            return this;
        }

        /**
         * Scheduler address: {@code host:port}, {@code :port}, or {@code unix:/path}.
         *
         * @param address where the scheduler is listening
         * @return {@code this}, for chaining
         */
        public Builder attach(String address) {
            this.address = address;
            return this;
        }

        /**
         * Jobs to run at once (default 1).
         *
         * @param slots the concurrency; values below 1 are raised to 1
         * @return {@code this}, for chaining
         */
        public Builder slots(int slots) {
            this.slots = Math.max(1, slots);
            return this;
        }

        /**
         * Shared secret, when the scheduler requires one.
         *
         * @param token the secret, or {@code null} when the scheduler needs none
         * @return {@code this}, for chaining
         */
        public Builder token(@Nullable String token) {
            this.token = token;
            return this;
        }

        /**
         * Identity announced to the scheduler (default: generated per process).
         *
         * @param executorId the id, or {@code null} to generate one
         * @return {@code this}, for chaining
         */
        public Builder executorId(@Nullable String executorId) {
            this.executorId = executorId;
            return this;
        }

        /**
         * How long to wait for the connection (default 10000ms).
         *
         * @param millis the connect timeout in milliseconds
         * @return {@code this}, for chaining
         */
        public Builder connectTimeoutMs(long millis) {
            this.connectTimeoutMs = millis;
            return this;
        }

        /**
         * How often to send a liveness heartbeat (default 5000ms).
         *
         * @param millis the interval in milliseconds
         * @return {@code this}, for chaining
         */
        public Builder heartbeatIntervalMs(long millis) {
            this.heartbeatIntervalMs = millis;
            return this;
        }

        /**
         * How long a drain waits for in-flight jobs (default 30000ms).
         *
         * @param millis the drain window in milliseconds
         * @return {@code this}, for chaining
         */
        public Builder shutdownDrainMs(long millis) {
            this.shutdownDrainMs = millis;
            return this;
        }

        /**
         * Payload serializer (default JSON). Must match the enqueuing side.
         *
         * @param serializer what decodes payloads and encodes results
         * @return {@code this}, for chaining
         */
        public Builder serializer(Serializer serializer) {
            this.serializer = serializer;
            return this;
        }

        /**
         * Middleware applied around every handler.
         *
         * @param middleware the chain, in the order it wraps the handler
         * @return {@code this}, for chaining
         */
        public Builder middleware(List<Middleware> middleware) {
            this.middleware = List.copyOf(middleware);
            return this;
        }

        /**
         * Named payload codecs, for tasks declaring {@code codecs}.
         *
         * @param codecs the codecs by name, resolved against each task's declared list
         * @return {@code this}, for chaining
         */
        public Builder codecs(Map<String, PayloadCodec> codecs) {
            this.codecs = Map.copyOf(codecs);
            return this;
        }

        /**
         * Injectable resources available to handlers.
         *
         * @param resources the runtime handlers resolve {@code Resources.use(...)} against
         * @return {@code this}, for chaining
         */
        public Builder resources(ResourceRuntime resources) {
            this.resources = resources;
            return this;
        }

        /**
         * Subscribe to a worker lifecycle or job event.
         *
         * @param name the event to subscribe to
         * @param listener called on the emitting thread
         * @return {@code this}, for chaining
         */
        public Builder on(EventName name, Consumer<FlexiQEvent> listener) {
            listeners.computeIfAbsent(name, key -> new ArrayList<>()).add(listener);
            return this;
        }

        /**
         * Task names this executor will advertise.
         *
         * @return the registered names; the scheduler dispatches nothing outside them
         */
        public List<String> tasks() {
            return List.copyOf(handlers.keySet());
        }

        /**
         * Dial the scheduler and start running jobs.
         *
         * <p>The handshake happens here, so a bad token or an unreachable
         * scheduler throws before any handler thread exists.
         *
         * @return the running executor, to be closed when the process stops
         */
        public Executor start() {
            if (address == null || address.isBlank()) {
                throw new IllegalStateException(
                        "no scheduler address: call attach(...) or set FLEXIQ_ATTACH (e.g. scheduler:7777)");
            }
            if (handlers.isEmpty()) {
                // The scheduler only dispatches task names an executor
                // advertises, so this would attach and then sit idle forever.
                throw new IllegalStateException("no handlers registered: call discover() or register(...)");
            }

            ExecutorService pool = Executors.newFixedThreadPool(slots);
            Emitter emitter = new Emitter();
            listeners.forEach((name, bound) -> bound.forEach(listener -> emitter.onEvent(name, listener)));

            // No QueueBackend: an executor reads no storage, which is the point
            // of the split. Job metadata and dashboard middleware toggles are
            // storage-backed and so unavailable here.
            WorkerDispatchBridge bridge =
                    new WorkerDispatchBridge(null, handlers, serializer, pool, emitter, middleware, resources, codecs);

            long handle;
            try {
                handle = NativeExecutor.attach(bridge, encodeOptions());
            } catch (RuntimeException e) {
                pool.shutdownNow();
                throw e;
            }
            JniExecutorControl control = new JniExecutorControl(handle);
            try {
                bridge.bind(control);
                emitter.emit(new WorkerEvent(EventName.WORKER_STARTED, List.of()));
                // Lease worker resources only after the attach succeeded, so a
                // refused handshake leaks nothing.
                resources.acquireWorker();
            } catch (RuntimeException e) {
                // The attach already succeeded, so the socket, the reader thread
                // and the native handle are live; a throwing listener or
                // resource factory must not strand them for the process's life.
                control.close();
                pool.shutdownNow();
                throw e;
            }
            return new Executor(control, pool, resources, emitter);
        }

        /** The attach options, as the JSON the native side parses. */
        private String encodeOptions() {
            Map<String, Object> options = new LinkedHashMap<>();
            options.put("address", address);
            options.put("tasks", tasks());
            options.put("slots", slots);
            if (token != null) {
                options.put("token", token);
            }
            if (executorId != null) {
                options.put("executorId", executorId);
            }
            if (connectTimeoutMs != null) {
                options.put("connectTimeoutMs", connectTimeoutMs);
            }
            if (heartbeatIntervalMs != null) {
                options.put("heartbeatIntervalMs", heartbeatIntervalMs);
            }
            if (shutdownDrainMs != null) {
                options.put("shutdownDrainMs", shutdownDrainMs);
            }
            try {
                return JSON.writeValueAsString(options);
            } catch (Exception e) {
                throw new IllegalStateException("failed to encode executor options", e);
            }
        }

        @SuppressWarnings("unchecked")
        private static org.byteveda.flexiq.task.TaskFunction<Object, Object> cast(Object function) {
            return (org.byteveda.flexiq.task.TaskFunction<Object, Object>) function;
        }
    }
}
